### DISCUSSION TITLE
Possible Optimizations

### DISCUSSION THREAD

**Discussion by azrahello** [filipstrand/mflux#550] [Ideas & Suggestions: Features]
Hi, would it be possible to implement an upscaler? Either using ControlNet like this: https://huggingface.co/jasperai/Flux.1-dev-Controlnet-Upscaler or something similar to https://github.com/ssitu/ComfyUI_UltimateSDUpscale? I find the ability to upscale images very useful. I was wondering if there’s a chance to implement it.
P.S. If you need a tester for your projects, I’m always available! 😉
Edit. 

Yesterday, I tested [this paper](https://arxiv.org/pdf/2411.19108) and [this repository](https://github.com/ali-vilab/TeaCache) in ComfyUI using FLUX1 dev in 8-bit GGUF. By following this process, I was able to almost halve the number of iterations and the time required, effectively making it faster than MLX. I was wondering if this could be useful and if it could be implemented. It would be amazing to generate images with FLUX in almost half the traditional time!

**Comment by filipstrand** [comment]
Hi @azrahello! Upscaling is interesting for sure and should be something we should support! Thanks for linking the ControlNet one, I will look into this to get a sense of how much work it would take to implement... Depending on a few personal factors, I might be able to spend more time than typical on MFLUX over the coming time so I am really happy with your feature suggestions here :)  

TeaCache looks awesome,  definitely worth investigating! On  their site, they have an example to get a sense of the quality degradation, and that looks really promising. 
<img width="1045" alt="Screenshot 2025-01-14 at 19 27 49" src="https://github.com/user-attachments/assets/b4d77988-855d-41df-8ea8-b1f93b2c7e87" />

Would you say with your own testing that you got similary good results, or do they vary a lot? Anyway, really cool!

**Comment by azrahello** [comment]
this is the output of comfy without teacache, 
Requested to load Flux
100%|███████████████████████████████████████████| 25/25 [01:17<00:00,  3.09s/it]
Prompt executed in 77.91 seconds
got prompt
this with teacache
Requested to load Flux
100%|███████████████████████████████████████████| 25/25 [00:40<00:00,  1.62s/it]
Prompt executed in 40.91 seconds  this with teacache
![ComfyUI_00135_](https://github.com/user-attachments/assets/423296fc-9c01-492f-b9f7-9ce539487afb)
this without teacache. 
![ComfyUI_00134_](https://github.com/user-attachments/assets/eec11079-935c-48d9-b4a4-02ed0d84251e)
with a value of 0.40.
 This is made with mflux. 
![image](https://github.com/user-attachments/assets/eed623a5-fda7-443a-a182-b176b27efc6a)
 time mflux-generate \
--prompt "Luxury food photograph" \
--model dev \
--steps 25 \
--seed 1010 \
--height 512 \
--width 768
Fetching 8 files: 100%|████████████████████████| 8/8 [00:00<00:00, 75573.05it/s]
Fetching 7 files: 100%|████████████████████████| 7/7 [00:00<00:00, 19277.83it/s]
100%|███████████████████████████████████████████| 25/25 [01:02<00:00,  2.51s/it]
mflux-generate --prompt "Luxury food photograph" --model dev --steps 25 --see  3,65s user 39,06s system 64% cpu 1:06,27 total

**Comment by azrahello** [comment]
Hi! Still in the same field, I also found this node: [Comfy-WaveSpeed](https://github.com/chengzeyi/Comfy-WaveSpeed?tab=readme-ov-file), which in the README refers to this: [ParaAttention](https://github.com/chengzeyi/ParaAttention). I’m not sure if they do the same thing, but while teacache works for me with GGUF, these others don’t.

Also, I saw that MLX Explore released this: “Support loading F8_E4M3 from safetensors”, which is a request I often see in nodes for loading the Flux model in ComfyUI (but it’s not usable because it requires too many resources). Maybe this could be useful for you? For me, it’s totally out of reach! :P

**Comment by filipstrand** [comment]
@azrahello Hi! Thanks for the pointers, I have not seen these before and your great suggestions are always welcome :) 

I remember I tried the tea-cache technique a while back but did not give it enough time to properly evaluate it and put it on hold for now. It also felt a bit "hackish" with a few magic numbers in the implementation that made me a bit suspicious..., but can be worth revisiting. 

For [First Block Cache](https://github.com/chengzeyi/ParaAttention/blob/main/doc/fastest_flux.md), it looks like they claim it is even simpler than TeaCache which sounds promising. I'll put this on the reading list and revisit after finishing up the 0.6.0 release!

**Comment by azrahello** [comment]
Hi, today I found this on Reddit: “https://www.reddit.com/r/StableDiffusion/s/VZHQWZWsc4”. It introduces a lossless compression method that should allow execution on machines with less RAM.
The links included in the article are:
	•	https://huggingface.co/DFloat11/FLUX.1-dev-DF11
	•	https://arxiv.org/abs/2504.11651
	•	https://github.com/LeanModels/DFloat11/tree/master/examples/flux.1
	•	https://huggingface.co/DFloat11/FLUX.1-schnell-DF11
	•	https://huggingface.co/DFloat11/FLUX.1-dev-DF11

I’m not sure if this could be applied to MLX or Apple hardware

**Comment by IonDen** [comment]
I built `mlx-teacache` since it's what this thread is about: it adds TeaCache to mflux as a standalone wrapper, not a fork. You register it like any other callback and it patches the predict step at runtime, caching a denoising step's transformer output and reusing it on the next step when the predicted change is small enough.

Measured on an M1 Max:
- FLUX.1-dev, 25 steps: 6 of 25 steps skipped, 1.46× wall-clock, SSIM ≥ 0.90 on the gate prompt.
- Non-distilled FLUX.2 Klein base, 50 steps + CFG: 1.23–1.36× depending on the variant.

Distilled 4–8 step Klein schedules don't get algorithmic skipping (adjacent steps differ too much for the gate to fire), so it's targeted at the longer schedules above.

Repo and per-variant notes: https://github.com/IonDen/mlx-teacache

**Comment by azrahello** [comment]
> I built `mlx-teacache` since it's what this thread is about: it adds TeaCache to mflux as a standalone wrapper, not a fork. You register it like any other callback and it patches the predict step at runtime, caching a denoising step's transformer output and reusing it on the next step when the predicted change is small enough.
> 
> Measured on an M1 Max:
> 
> * FLUX.1-dev, 25 steps: 6 of 25 steps skipped, 1.46× wall-clock, SSIM ≥ 0.90 on the gate prompt.
> * Non-distilled FLUX.2 Klein base, 50 steps + CFG: 1.23–1.36× depending on the variant.
> 
> Distilled 4–8 step Klein schedules don't get algorithmic skipping (adjacent steps differ too much for the gate to fire), so it's targeted at the longer schedules above.
> 
> Repo and per-variant notes: https://github.com/IonDen/mlx-teacache

Hello! I've seen your work and it's really piqued my interest—what are you working on right now? :D

I have to say, I've been using your tools, especially mlx-taef (in PR). Currently, I'm experimenting with ComfyUI's interface and trying to make omlx, mflux, and mlx-video coexist  in  order to build complex workflows that I wouldn't know how to handle via CLI.

It's funny, and I think something equivalent and versatile is still missing on Apple.

<img width="1717" height="1019" alt="Image" src="https://github.com/user-attachments/assets/5aa0ef24-9faf-4ec1-b96f-cfa80501911d" />
<img width="655" height="830" alt="Image" src="https://github.com/user-attachments/assets/2df794f9-8ecd-47a0-b80e-40a370460bd8" />

**Comment by IonDen** [comment]
Hi, thanks @azrahello I also think Apple has great potential on AI landscape and can be a strong competitor to Nvidia, but so many tools are missing now.

**Comment by ianscrivener** [comment]
@IonDen , what tools do you feel are missing?

Personally, when I feel I am missing tools... it often turns out that I can actually do it on a Mac (I use MFlux, DrawThings & ComfyUI)... and the actual issue a block/challenge on the creativity side of my process. Granted a H100 will do stuff faster than my Mac Studio Max.

**Comment by ianscrivener** [comment]
Is Teacache actually worthwhile?
I'm being genuine here. I did extensive tests with Teacache on DrawThings with Flux.1 Base - comparing it to Flux.1 Dev. I was very disappointed - Flux.1 Base + teacahe was still much slower than Flux.1 Dev - and the same (Flux.1 meh) image quality. Broadly the DrawThings Discord community seemed to share my assessment of the usefulness of Teacache for Flux.1. 

Is Teacache for Flux.2 Klein Base better + faster than Flux.2 Klein Dev? Test images and benchmark times would be greatly appreciated!!

**Comment by IonDen** [comment]
@ianscrivener I have a benchmark and test results. Check it here: https://github.com/IonDen/mlx-teacache/blob/main/COMPARISON.md

**Comment by IonDen** [comment]
Of course you have all major ones on Mac. But still Metal ecosystem is much less mature than CUDA ecosystem.

**Comment by ianscrivener** [comment]
> [@ianscrivener](https://github.com/ianscrivener) I have a benchmark and test results. Check it here: https://github.com/IonDen/mlx-teacache/blob/main/COMPARISON.md

thanks **@IonDen**,  
The comparison that I'm more interested in would be;
 - **Flux.2 Klein** versus **Flux.2 Klein Base teacache**

ie is there an compelling reason to use base + teacache... or just keep using Dev as the working model for generations?
Sure, base is required for finetunes...

**Reply by IonDen** [reply]
@ianscrivener, better late than never. No, Base + TeaCache does not beat the distilled Klein on speed, and it is not close.

The arithmetic: distilled Klein runs 4 transformer passes in total. Klein Base at 50 steps with CFG runs about 100, two per step. TeaCache skips 8 of the 48 gated steps on the 4B and 13 on the 9B, which leaves Base with around 20 times more transformer work than distilled.

Measured on an M1 Max 32 GB at q4: Klein Base 4B, 768x1024, 50 steps, guidance 4: 573 s plain, 486 s with TeaCache (1.18x). Klein Base 9B, 512x512: 521 s plain, 379 s with TeaCache (1.37x, SSIM 0.986 against plain). Images and method: https://github.com/IonDen/mlx-teacache/blob/main/COMPARISON.md

So the reason to run Base is not speed. It is guidance, negative prompts and fine-tunes, and TeaCache makes that path 15 to 27 percent cheaper. For everyday generation, keep the distilled model.

I am writing a small script that runs all three (distilled, Base, Base + TeaCache) on one prompt and seed and prints a table. I can only run it at q4 here; if anyone with a 64 GB Mac wants to run it at q8 or bf16, I will post the link.

**Reply by IonDen** [reply]
As promised, @ianscrivener: the three-way script is merged. It runs distilled, Base, and Base + TeaCache on one prompt and seed and prints a table. The "Distilled Klein vs Base Klein" section in COMPARISON.md has the tables and images: https://github.com/IonDen/mlx-teacache/blob/main/COMPARISON.md

At q4 on an M1 Max, Base + TeaCache is 1.20× over Base on the 4B (8 of 48 steps skipped, SSIM 0.93 to plain Base) and 1.42× on the 9B (14 skipped, SSIM 0.98). Distilled stays roughly 20× faster than either, so it is still the everyday model. Base is for when you need guidance, negative prompts, or a fine-tune, and the wrapper makes that path 15 to 30% cheaper.

I can only measure q4 here. On a 64 GB Mac, `uv run python scripts/bench_klein_base_vs_distilled.py --size 9b --quantize 8 --reps 3` prints the same table at q8, and I will fold your numbers in.

**Reply by ianscrivener** [reply]
a 15% to 27% speed up is impressive!!

Yes agree: base wins on neg prompts, guidance & fine tuneability...

Pleasantly surprised to see teacache is reducing peak memory by so much... though the WHY of that is above my pay grade!

**Reply by IonDen** [reply]
Hi @ianscrivener, that memory drop isn't really TeaCache earning its keep. It's a side effect of how the integration runs the model. Plain mflux compiles the transformer's forward pass, and the compiler holds extra working memory for that compiled graph. Our wrapper runs the model eagerly instead, so that working memory is never reserved and the peak comes out lower.

The step-skipping is what buys the speed, not the memory. Peak memory is the highest single point of use across a whole run, so skipping some steps can't lower it: one full step still hits the same high mark. Our own bench shows it plainly: a variant that skips zero steps still lands at the same low peak.

Numbers so far are all 32 GB M1 Max at q4. If you get a chance to run the [three-way script](https://github.com/IonDen/mlx-teacache/blob/main/scripts/bench_klein_base_vs_distilled.py) on your 64 GB Mac at q8 or bf16, that would be genuinely useful: it'd show whether the memory and speed hold at higher precision, on hardware I can't test here.

**Reply by ianscrivener** [reply]
Yes, There is a lot of room for optimization in MFlux. 

I put the effort into build a comprehensive model library of pre-quantized MFlux format models... some people just do not have the available memory or bandwidth to pull down big upstream bf16 models and quantize them locally.

**[MFlux-Community](https://huggingface.co/mflux-community) on HF now has over 200 MFlux models**
